crates/ide-assists/src/handlers/bool_to_enum.rs

```diff
@@ -430,18 +459,48 @@ fn add_enum_def(
         .any(|module| module.nearest_non_block_module(ctx.db()) != *target_module);
     let enum_def = make_bool_enum(make_enum_pub);
 
-    let indent = IndentLevel::from_node(&target_node);
+    let insert_before = node_to_insert_before(target_node);
+    let indent = IndentLevel::from_node(&insert_before);
     enum_def.reindent_to(indent);
 
     ted::insert_all(
-        ted::Position::before(&edit.make_syntax_mut(target_node)),
+        ted::Position::before(&edit.make_syntax_mut(insert_before)),
         vec![
             enum_def.syntax().clone().into(),
             make::tokens::whitespace(&format!("\n\n{indent}")).into(),
         ],
     );
 }
 
+/// Finds where to put the new enum definition, at the nearest module or at top-level.
+fn node_to_insert_before(mut target_node: SyntaxNode) -> SyntaxNode {
+    let mut ancestors = target_node.ancestors();
+
+    while let Some(ancestor) = ancestors.next() {
+        match_ast! {
+            match ancestor {
+                ast::Item(item) => {
+                    if item
+                        .syntax()
+                        .parent()
+                        .and_then(|item_list| item_list.parent())
+                        .and_then(ast::Module::cast)
+                        .is_some()
+                    {
+                        return ancestor;
+                    }
+                },
+                ast::SourceFile(_) => break,
+                _ => (),
+            }
+        }
+
+        target_node = ancestor;
+    }
+
+    target_node
+}
+
```

Comment:
This can be simplified to
```suggestion
/// Finds where to put the new enum definition, at the nearest module or at top-level.
fn node_to_insert_before(target_node: SyntaxNode) -> SyntaxNode {
    target_node.ancestors().find(|it| matches!(it.kind(), SyntaxKind::Module | SyntaxKind::SourceFile).unwrap_or(target_node)
}

```

Reply:
Oh sorry that wes me spacing out, in that case we can do
```rs
target_node.ancestors().take_while(|it| !matches!(it.kind(), SyntaxKind::Module | SyntaxKind::SourceFile).last().unwrap_or(target_node)
```

Reply:
No problem! I think this works for the source file case but in the module case we want the second to last node since the last is an `ItemList` and inserting it before the `ItemList` would be between the module name and the left brace.

Reply:
Right, then my snippet only needs a slight adjustment by adding a filter
```rs
target_node.ancestors().take_while(|it| !matches!(it.kind(), SyntaxKind::Module | SyntaxKind::SourceFile).filter(|it| ast::Item::can_cast(it.kind())).last().unwrap_or(target_node)
```

Though I feel like there has to be some assist that does something similar here, I'd be surprised if not but I can't think of any of the top oh my head right now.

Reply:
Ah I see okay, that works and is is much cleaner, thank you! 

I thought that would be the case as well, a few that I looked at were `extract_function`, which has `node_to_insert_after` and is what I based this one on but seems to be a bit more complex and `generate_function` which has `next_space_for_fn_in_module` and similar functions but is based on text ranges and also seem more complex. Do you think it would be better to try to reuse these?

Reply:
nah this is simple enough so it should be fine